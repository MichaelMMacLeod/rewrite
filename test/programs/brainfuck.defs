// Implements a brainfuck interpreter. See: https://en.wikipedia.org/wiki/Brainfuck
// Instead of inputting/outputting chars, this receives/produces Peano integers, since
// this is the easiest way to implement addition/subtraction in this language.
// See: https://en.wikipedia.org/wiki/Peano_axioms.

(def (bf (input = $in) $x ..)
     (program
       (input = $in)
       (output = ())
       (mem = (() (0)))
       (cmd = (read (()) $x ..))))

(def (read ($c ..) [ $x ..)
  (read (() $c ..) $x ..))

(def (read ($c0 ($c ..) $c1 ..) ] $x ..)
  (read (($c .. $c0) $c1 ..) $x ..))

(def (read (($c0 ..) $c ..) . $x ..)
  (read (($c0 .. .) $c ..) $x ..))

(def (read (($c0 ..) $c ..) , $x ..)
  (read (($c0 .. ,) $c ..) $x ..))

(def (read (($c0 ..) $c ..) + $x ..)
  (read (($c0 .. +) $c ..) $x ..))

(def (read (($c0 ..) $c ..) - $x ..)
  (read (($c0 .. -) $c ..) $x ..))

(def (read (($c0 ..) $c ..) > $x ..)
  (read (($c0 .. >) $c ..) $x ..))

(def (read (($c0 ..) $c ..) < $x ..)
  (read (($c0 .. <) $c ..) $x ..))

(def (read (($c ..)))
  (() ($c ..)))

(def (program
       (input = $in)
       (output = $out)
       (mem = $d))
     (result = $out))

(def (program
       (input = $in)
       (output = $out)
       (mem = $mem)
       (cmd = (($i0 ..) ()))
       $stack ..)
     (program
       (input = $in)
       (output = $out)
       (mem = $mem)
       $stack ..))

(def (program
       (input = $in)
       (output = $out)
       (mem = (($db ..) ($d $d0 $da ..)))
       (cmd = (($ib ..) (> $ia ..)))
       $stack ..)
     (program
       (input = $in)
       (output = $out)
       (mem = (($db .. $d) ($d0 $da ..)))
       (cmd = (($ib .. >) ($ia ..)))
       $stack ..))

(def (program
       (input = $in)
       (output = $out)
       (mem = (($db ..) ($d)))
       (cmd = (($ib ..) (> $ia ..)))
       $stack ..)
     (program
       (input = $in)
       (output = $out)
       (mem = (($db .. $d) (0)))
       (cmd = (($ib .. >) ($ia ..)))
       $stack ..))

(def (program
       (input = $in)
       (output = $out)
       (mem = (($db .. $d0) ($d $da ..)))
       (cmd = (($ib ..) (< $ia ..)))
       $stack ..)
     (program
       (input = $in)
       (output = $out)
       (mem = (($db ..) ($d0 $d $da ..)))
       (cmd = (($ib .. <) ($ia ..)))
       $stack ..))

(def (program
       (input = $in)
       (output = $out)
       (mem = (($db ..) ($d $da ..)))
       (cmd = (($ib ..) (+ $ia ..)))
       $stack ..)
     (program
       (input = $in)
       (output = $out)
       (mem = (($db ..) ((S $d) $da ..)))
       (cmd = (($ib .. +) ($ia ..)))
       $stack ..))

(def (program
       (input = $in)
       (output = $out)
       (mem = (($db ..) ((S $d) $da ..)))
       (cmd = (($ib ..) (- $ia ..)))
       $stack ..)
     (program
       (input = $in)
       (output = $out)
       (mem = (($db ..) ($d $da ..)))
       (cmd = (($ib .. -) ($ia ..)))
       $stack ..))

(def (program
       (input = $in)
       (output = ($out ..))
       (mem = (($db ..) ($d $da ..)))
       (cmd = (($ib ..) (. $ia ..)))
       $stack ..)
     (program
       (input = $in)
       (output = ($out .. $d))
       (mem = (($db ..) ($d $da ..)))
       (cmd = (($ib .. .) ($ia ..)))
       $stack ..))

(def (program
       (input = ($in0 $in ..))
       (output = $out)
       (mem = (($db ..) ($d $da ..)))
       (cmd = (($ib ..) (, $ia ..)))
       $stack ..)
     (program
       (input = ($in ..))
       (output = $out)
       (mem = (($db ..) ($in0 $da ..)))
       (cmd = (($ib .. ,) ($ia ..)))
       $stack ..))

(def (program
       (input = ())
       (output = $out)
       (mem = (($db ..) ($d $da ..)))
       (cmd = (($ib ..) (, $ia ..)))
       $stack ..)
     (program
       (input = ())
       (output = $out)
       (mem = (($db ..) (0 $da ..)))
       (cmd = (($ib .. ,) ($ia ..)))
       $stack ..))

(def (program
       (input = $in)
       (output = $out)
       (mem = (($db ..) (0 $da ..)))
       (cmd = (($ib ..) (($ii ..) $ia ..)))
       $stack ..)
     (program
       (input = $in)
       (output = $out)
       (mem = (($db ..) (0 $da ..)))
       (cmd = (($ib .. ($ii ..)) ($ia ..)))
       $stack ..))

(def (program
       (input = $in)
       (output = $out)
       (mem = (($db ..) ((S $d) $da ..)))
       (cmd = (($ib ..) (($ii ..) $ia ..)))
       $stack ..)
     (program
       (input = $in)
       (output = $out)
       (mem = (($db ..) ((S $d) $da ..)))
       (cmd = (() ($ii ..)))
       (cmd = (($ib ..) (($ii ..) $ia ..)))
       $stack ..))