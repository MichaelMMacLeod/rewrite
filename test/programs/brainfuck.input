// Sorts the list [3, 2, 1] using quicksort. I copied
// the brainfuck-quicksort program from here:
// https://www.codingame.com/playgrounds/50516/brainfuck-part-9---sort-arrays-with-bubble-and-quicksort/quicksort

(bf (input = ((S (S (S 0))) (S (S 0)) (S 0)))
  > , [ >
  + [ - > + < ] , ] > [ - > > > > > > > > > > > > >
  > > + < < < < < < < < < < < < < < < ] > > > > > >
  > > > > > > > > + [ [ - < + < + > > ] < [ - > + <
  ] < - [ - < < [ < ] < < < < < < < < < < < < [ - >
  > > > > > > > > > > > + < < < < < < < < < < < < ]
  > > > > > > > > > > > > [ > ] > ] > > > [ - < < +
  < + > > > ] < < [ - > > + < < ] < [ - < < [ < ] <
  < < < < < < < < + > > > > > > > > > > [ > ] > ] <
  < [ < ] < < < < < < < < < [ - < < [ < ] < [ - > +
  < ] > [ > ] > ] < < [ - > > > > + < < < < ] < [ [
  - > + > + < < ] > > [ - < < + > > ] < [ - > > [ >
  ] > > > > + < < < < < [ < ] < ] > > [ > ] > > [ -
  > + < < < + > > ] > [ - < + > ] > [ - < + > ] < <
  > > + < < [ - > [ - > ] > [ < < [ - ] > > - > > ]
  < < + < < ] > [ [ - ] < + > ] < [ - > + < ] < < [
  - > > + < < ] > > > [ - > - < < < < < [ < ] < < [
  - > > > [ > ] > > > > > > > > + < < < < < < < < <
  [ < ] < < ] > > > [ > ] > > > > > > > + [ - < + >
  ] < < < < < < < < [ < ] > [ [ - < + > ] > ] > [ -
  < + > ] > [ - < + > ] ] > [ - < < < + < < [ < ] <
  < [ - > > + < < ] > > [ > ] > > > > ] < < < < < [
  < ] < < ] > > > [ > ] > > > > > > > [ [ - < + < +
  > > ] < [ - > + < ] > [ - > > [ > ] > [ > ] > > +
  < < < [ < ] < < [ < ] < ] > > [ > ] > [ > ] > > >
  [ - < < < + > > > ] < < < [ - > > > + [ > ] > + <
  < [ < ] < ] > > [ - < + > ] < [ - > > [ > ] > > +
  < < < [ < ] < ] > > [ > ] > [ [ - < + > ] > ] < <
  [ < ] < < < [ < ] < [ < ] < ] < < < < < < [ [ - >
  > > > > > > > [ > ] > [ > ] > > > [ > ] > > + < <
  < [ < ] < < < [ < ] < [ < ] < < < < < < < ] > > >
  > [ - > > > > [ > ] > [ > ] > + < < [ < ] < [ < ]
  < < < ] > > > > [ > ] > [ > ] > > > [ - < + < + >
  > ] < [ - > + < ] < + [ - > > [ > ] > + < < [ < ]
  < ] > > [ > ] > [ [ - < + > ] > ] < < [ < ] < < <
  [ < ] < [ < ] < < < < < < < ] < < [ < ] > [ [ - <
  < < + > > > ] > ] > > [ - < < < < < + > > > > > ]
  > > > [ - ] > > > > [ [ - < < < < < < < < < < < +
  > > > > > > > > > > > ] > ] > [ [ - < < < < < < <
  < < < < < + > > > > > > > > > > > > ] > ] > > > [
  - ] > [ - ] > [ [ - < < + > > ] > ] < < < [ < ] >
  ] < < < < < < < < < < < < < < < < [ < ] > [ . > ]
)
