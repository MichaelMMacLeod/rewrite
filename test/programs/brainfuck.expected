(result = ((S 0) (S (S 0)) (S (S (S 0)))))
